New SMB/WMI Module BitLocker

termanix opened the pull request:
Checking BitLocker status on all drives.

![image](https://github.com/Pennyw0rth/NetExec/assets/50464194/9a137f60-9e8d-4580-aba7-8e2cf3c89072)

Adamkadaban commented:
Would it perhaps be better to run this via WMI? Seems slightly more op-sec safe.

I haven't tested, but it seems possible: https://4sysops.com/archives/check-the-bitlocker-status-of-all-pcs-in-the-network/#rtoc-3

termanix commented:
> Would it perhaps be better to run this via WMI? Seems slightly more op-sec safe.
> 
> I haven't tested, but it seems possible: https://4sysops.com/archives/check-the-bitlocker-status-of-all-pcs-in-the-network/#rtoc-3

Test's an great idea, I'm testing now, I will share results manually. Thank you.

termanix commented:
It works @Adamkadaban I can also adapt it to this method. Let me edit it.

![image](https://github.com/Pennyw0rth/NetExec/assets/50464194/e5393e24-7b2c-4bb2-b58d-0f656823c83f)

Adamkadaban commented:
Just checked and it seems WMI works. Also produces less security events
![image](https://github.com/Pennyw0rth/NetExec/assets/34610663/2028febe-84b4-4ab9-987b-a99bc1aa1683)


I think this should work for doing it through WMI:

```python
iInterface = dcom.CoCreateInstanceEx(wmi.CLSID_WbemLevel1Login,wmi.IID_IWbemLevel1Login)
iWbemLevel1Login = wmi.IWbemLevel1Login(iInterface)

bitlockerNamespace = "root\\CIMv2\\Security\\MicrosoftVolumeEncryption"
try:
    iWbemServices= iWbemLevel1Login.NTLMLogin(bitlockerNamespace, NULL, NULL)
except Exception as e:
    if str(e).find("WBEM_E_INVALID_NAMESPACE") >= 0:
        iWbemLevel1Login.RemRelease()
        dcom.disconnect()
    else:
        nxc_logger.debug(str(e))

iWbemServices.get_dce_rpc().set_auth_level(RPC_C_AUTHN_LEVEL_PKT_PRIVACY)

classQuery = "SELECT DriveLetter, ConversionStatus, ProtectionStatus, EncryptionMethod FROM Win32_EncryptableVolume"

try:
    iEnumWbemClassObject = iWbemServices.ExecQuery(classQuery)
    iWbemClassObject = iEnumWbemClassObject.Next(0xffffffff,1)[0]
    queryResult = iWbemClassObject.getProperties()
    # parse and print data from here
    iEnumWbemClassObject.RemRelease()
except Exception as e:
    if str(e).find("WBEM_E_INVALID_CLASS") >= 0:
        iWbemLevel1Login.RemRelease()
        iWbemServices.RemRelease()
        dcom.disconnect()
    print(str(e))
    nxc_logger.debug(str(e))

iWbemLevel1Login.RemRelease()
iWbemServices.RemRelease()
dcom.disconnect()
```

termanix commented:
> @termanix Sorry, I should've been clearer. I meant through the WMI protocol rather than through powershell
> 
> Just checked with wmiquery and it seems to work. Also produces less security events
> 
> I think this should work for doing it through WMI:
> 
> ```python
> iInterface = dcom.CoCreateInstanceEx(wmi.CLSID_WbemLevel1Login,wmi.IID_IWbemLevel1Login)
> iWbemLevel1Login = wmi.IWbemLevel1Login(iInterface)
> 
> bitlockerNamespace = "root\\CIMv2\\Security\\MicrosoftVolumeEncryption"
> try:
>     iWbemServices= iWbemLevel1Login.NTLMLogin(bitlockerNamespace, NULL, NULL)
> except Exception as e:
>     if str(e).find("WBEM_E_INVALID_NAMESPACE") >= 0:
>         iWbemLevel1Login.RemRelease()
>         dcom.disconnect()
>     else:
>         nxc_logger.debug(str(e))
> 
> iWbemServices.get_dce_rpc().set_auth_level(RPC_C_AUTHN_LEVEL_PKT_PRIVACY)
> 
> classQuery = "SELECT DriveLetter, ConversionStatus, ProtectionStatus, EncryptionMethod FROM Win32_EncryptableVolume"
> 
> try:
>     iEnumWbemClassObject = iWbemServices.ExecQuery(classQuery)
>     iWbemClassObject = iEnumWbemClassObject.Next(0xffffffff,1)[0]
>     queryResult = iWbemClassObject.getProperties()
>     # parse and print data from here
>     iEnumWbemClassObject.RemRelease()
> except Exception as e:
>     if str(e).find("WBEM_E_INVALID_CLASS") >= 0:
>         iWbemLevel1Login.RemRelease()
>         iWbemServices.RemRelease()
>         dcom.disconnect()
>     print(str(e))
>     nxc_logger.debug(str(e))
> 
> iWbemLevel1Login.RemRelease()
> iWbemServices.RemRelease()
> dcom.disconnect()
> ```

Oh okay, I didn't test it with wmi protocol yet. I was just be quick for get result. 

Using WMI is your idea and I don't want to steal your idea :D I can add it for WMI if it's okay to you.

Adamkadaban commented:
> I can add it for WMI if it's okay to you.

Yes, feel free. Add me as a co-author to the commit if you so please ;)

Marshall-Hallenbeck commented:
You could check `exec-method` and do SMB if it's smbexec, or WMI if it's wmiexec, and also add this to the `WMI` protocol and run it straight through WMI.

termanix commented:
> > I can add it for WMI if it's okay to you.
> 
> Yes, feel free. Add me as a co-author to the commit if you so please ;)

Ofc. I will add it for WMI too then. If you joined discord can you contact me? We can implement it together.

termanix commented:
It can be review now. I just cant pwned on Win 11 machine while using WMI protocol. But SMB works.
![image](https://github.com/Pennyw0rth/NetExec/assets/50464194/790ec88a-e16d-4969-b62d-6aa9246634f9)

Marshall-Hallenbeck commented:
@termanix can you add this module to the e2e tests file?

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+
+        try:
+            # Executing the PowerShell command to get BitLocker volumes status.
+            check_bitlocker_command_str_output = self.connection.execute(check_bitlocker_command_str, True)
```
You should be able to use `ps_execute()` for SMB, but for WMI it looks like you'll have to pass in the "powershell.exe..." command, otherwise your command ends up looking like `cmd.exe /Q /c powershell.exe "Get-BitLockerVolume | Select-Object MountPoint, EncryptionMethod, ProtectionStatus" 1> \Windows\Temp\jVuCDq 2>&1` during SMB

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+            
+            if "'Get-BitLockerVolume' is not recognized" in check_bitlocker_command_str_output:
+                self.context.log.fail("BitLockerVolume not found on target.")
+                sys.exit(1)
```
This shouldn't fail 1 - the command properly executed, but BitLockerVolume just wasn't found on the host. This will make the OS think that there was a programmatic failure in the code, which there wasn't.

termanix replied:
Done, Thank you ^^

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+            lines = check_bitlocker_command_str_output.strip().split("\n")
+
+            # Getting data lines.
+            data_lines = lines[2:]
```
You should be able to just add [2:] to line 54 above to reduce the amount of variables you are creating

termanix replied:
Done, thank you.

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+
+            # Analyzing data lines.
+            for line in data_lines:
+                
```
Remove unnecessary space

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+            # Analyzing data lines.
+            for line in data_lines:
+                
+                parts = re.split(r"\s{2,}", line.strip())  # Stripping spaces and splitting the line.
```
You should be able to do `MountPoint, EncryptionMethod, protection_status = re.split(r"\s{2,}", line.strip())`

termanix replied:
Done, thank you.

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+                        if iWbemClassObject[0].ProtectionStatus == 1:
+                            self.context.log.highlight(f"BitLocker is enabled on drive {iWbemClassObject[0].DriveLetter} (Encryption Method: {encryptionTypeMapping.get(encryptionMethod, 'Unknown')})")
+                        else:
+                            assert (encryptionMethod == 0)  # Should be 0 if disabled
```
Why is this not an if statement?

termanix replied:
Actually, @Adamkadaban wrote that part. I changed it as if statement.

Marshall-Hallenbeck commented on nxc/modules/bitlocker.py:
```diff
@@ -0,0 +1,138 @@
+            )
+            
+            try:
+
```
Remove unnecessary space

termanix replied:
Done, thank you.

termanix reviewed (commented):
Update e2e_commands.txt

termanix commented:
Updated @Marshall-Hallenbeck , It can be review.

quahac commented:
First of all, very nice module!  
Win10/11 is okay. However, Windows Server 2008 DC will raise an error on SMB:
![image](https://github.com/Pennyw0rth/NetExec/assets/49560894/20864a8c-6ead-48f5-9a1f-868f60bb29a9)

My approach is to check services instead of calling functions, for AV and Legacy. I have been using this method for a long time, and I like it. Ensure error handling (on endpoint) for this sneaky approach `... -EA SilentlyContinue ... `. Example:
![image](https://github.com/Pennyw0rth/NetExec/assets/49560894/bfc97589-c331-4db0-afc4-4465f7bf2d3f)

termanix commented:
> First of all, very nice module! Win10/11 is okay. However, Windows Server 2008 DC will raise an error on SMB: ![image](https://private-user-images.githubusercontent.com/49560894/331161144-20864a8c-6ead-48f5-9a1f-868f60bb29a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NTcwNDMsIm5iZiI6MTcxNTg1Njc0MywicGF0aCI6Ii80OTU2MDg5NC8zMzExNjExNDQtMjA4NjRhOGMtNmVhZC00OGY1LTlhMWYtODY4ZjYwYmIyOWE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEwNTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjYzE0NDc0ZTY3YzdkZjAzYzlhZjRiMTZhNGE4NWE1MzE0MTU5MmYxMmI3ZTEyNTg4YjRjNjcxMGMyZjcyZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aRWrPR7vJFzGrFFhkwBsBYMLzbZn3XiqMV8ukOLJEXM)
> 
> My approach is to check services instead of calling functions, for AV and Legacy. I have been using this method for a long time, and I like it. Ensure error handling (on endpoint) for this sneaky approach `... -EA SilentlyContinue ... `. Example: ![image](https://private-user-images.githubusercontent.com/49560894/331161886-bfc97589-c331-4db0-afc4-4465f7bf2d3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NTcwNDMsIm5iZiI6MTcxNTg1Njc0MywicGF0aCI6Ii80OTU2MDg5NC8zMzExNjE4ODYtYmZjOTc1ODktYzMzMS00ZGIwLWFmYzQtNDQ2NWY3YmYyZDNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEwNTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiZTQ5YzFiNWQ5Yzk4NzQyZDc5MTRmM2VjMjM1ZmFiNzZkYzgyOTYwMWYxZjJkM2YzN2JmMTQxMGYxOThkNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U9Imfzg2QgYxQX5PsGvR1C4ksB6KFOqplfD5_9f5x6w)

Hi, firstly thank you for your good comment ^^, I actually did it on older NetExec powershell. Now I fixed it again, i forgot it, thank you for reminding. Also thank you for suggestion. I added ``-EA SilentlyContinue`` for now, let's se how does it work.

termanix commented:
> @termanix , I see you're coding now. My mistake; I told you to use the -ErrorAction SilentlyContinue approach in my situation. However, since the function doesn't exist yet, you cannot call -EA, as it will also raise an error. You have to try the function and catch the error: `try{Get-BitLockerVolume | Select-Object MountPoint, EncryptionMethod, ProtectionStatus}catch{}`
> 
> ![image](https://private-user-images.githubusercontent.com/49560894/331200442-c5e7232e-18d3-495b-bd40-bbc69d57727b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NjUyNzUsIm5iZiI6MTcxNTg2NDk3NSwicGF0aCI6Ii80OTU2MDg5NC8zMzEyMDA0NDItYzVlNzIzMmUtMThkMy00OTViLWJkNDAtYmJjNjlkNTc3MjdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEzMDkzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliZWVmNTk0YTg4MjA5OGZhYTI4NjRiM2M3MjYyMWEwN2Y5MmI1ZGJjYTNhMGJjMTNlMTVjNWYyMjM3YmMxNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SrWa4CwK83ZnzoBviq-lKQEdlpXx_PM7cqVriazDKj4)

Oh okay got it. But about try catch, I just want to show it on NetExec output if system has not get-bitlocker. Code checks it in output. But I got it what you mean.

NeffIsBack reviewed (approved):
Changed command execution to cmd as this is more reliable in my experience (also in my tests). Also fixed some bugs with parsing the output.

Detecting Bitlocker on my laptop but not on the DC. LGTM now:
![image](https://github.com/Pennyw0rth/NetExec/assets/61382599/4d0e329a-6245-420d-851c-0d1391df1fa4)

NeffIsBack commented:
Can't test wmi on my notebook as it is not domain joined, but the negative detectin on the dc seems to work:
![image](https://github.com/Pennyw0rth/NetExec/assets/61382599/6cd1ab45-481e-4637-bb6b-b541e067d8f2)

@Marshall-Hallenbeck also saw your command regarding ps_execute. Changed it for now cause it fails on my most of the time and also is flagged pretty often by AVs (for example even win defender yelled at me lol). If that's fine for you please also approve (you have a pending change request)

Marshall-Hallenbeck reviewed (approved):
@NeffIsBack sounds good! Looks like my other changes were fixed

NeffIsBack commented:
Great! I'll merge🚀
